src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Windows.cs

```diff
@@ -0,0 +1,330 @@
+                }
+                else
+                {
+                    IPEndPoint ep = new IPEndPoint(address, 0);
+                    Internals.SocketAddress remoteAddr = IPEndPointExtensions.Serialize(ep);
+                    byte[] sourceAddr = new byte[28];
+
+                    SafeWaitHandle pingEventSafeWaitHandle = _pingEvent.GetSafeWaitHandle();
+                    error = (int)Interop.IpHlpApi.Icmp6SendEcho2(
+                        _handlePingV6,
+                        pingEventSafeWaitHandle,
+                        IntPtr.Zero,
+                        IntPtr.Zero,
+                        sourceAddr,
+                        remoteAddr.Buffer,
+                        _requestBuffer,
+                        (ushort)buffer.Length,
+                        ref ipOptions,
+                        _replyBuffer,
+                        MaxUdpPacket,
+                        (uint)timeout);
+                }
+            }
+            catch
+            {
+                UnregisterWaitHandle();
+                throw;
+            }
+
+            if (error == 0)
+            {
+                error = Marshal.GetLastWin32Error();
+
+                // Only skip Async IO Pending error value.
+                if (error != Interop.IpHlpApi.ERROR_IO_PENDING)
+                {
+                    // Cleanup.
+                    FreeUnmanagedStructures();
+                    UnregisterWaitHandle();
+
+                    throw new Win32Exception(error);
+                }
+            }
+        }
+
+        private void InternalDisposeCore()
+        {
+            if (_handlePingV4 != null)
+            {
+                _handlePingV4.Dispose();
+                _handlePingV4 = null;
+            }
+
+            if (_handlePingV6 != null)
+            {
+                _handlePingV6.Dispose();
+                _handlePingV6 = null;
+            }
+
+            UnregisterWaitHandle();
+
+            if (_pingEvent != null)
+            {
+                _pingEvent.Dispose();
+                _pingEvent = null;
+            }
+
+            if (_replyBuffer != null)
+            {
+                _replyBuffer.Dispose();
+                _replyBuffer = null;
+            }
+        }
+
+        static partial void InitializeSockets()
+        {
+            if (!Volatile.Read(ref s_socketInitialized))
+            {
+                lock (s_socketInitializationLock)
+                {
+                    if (!s_socketInitialized)
+                    {
+                        // Ensure that WSAStartup has been called once per process.  
+                        // The System.Net.NameResolution contract is responsible with the initialization.
+                        Dns.GetHostName();
+
+                        // Cache some settings locally.
+                        s_socketInitialized = true;
+                    }
+                }
+            }
+        }
+
+        // -----------------------------
```

Comment:
Instead of these comments, it would be better to follow the pattern that other System.Net PALs used:
* Ping.cs will contain the partial class for common code
* Ping.Windows.cs contains the Windows specific implementation
* Ping.Unix.cs contains the *NIX specific implementation